pkg/square/builder.go

```diff
@@ -0,0 +1,238 @@
+		}
+
+		// record the starting share index of the blob in the PFB that paid for it
+		c.pfbs[element.pfbIndex].ShareIndexes[element.blobIndex] = uint32(cursor)
+		// If this is not the first blob, we add padding by writing padded shares to the previous blob
+		// (which could be of a different namespace)
+		if i > 0 {
+			if err := blobWriter.WriteNamespacedPaddedShares(padding); err != nil {
+				return nil, fmt.Errorf("writing padding into sparse shares: %w", err)
+			}
+		}
+		// Finally write the blob itself
+		if err := blobWriter.Write(element.blob); err != nil {
+			return nil, fmt.Errorf("writing blob into sparse shares: %w", err)
+		}
+		// increment the cursor by the size of the blob
+		cursor += element.shareSize
+		endOfLastBlob = cursor
+	}
+
+	// write all the pay for blob transactions into compact shares. We need to do this after allocating the blobs to their
+	// appropriate shares as the starting index of each blob needs to be included in the PFB transaction
+	pfbTxWriter := shares.NewCompactShareSplitter(namespace.PayForBlobNamespace, appconsts.ShareVersionZero)
+	for _, iw := range c.pfbs {
+		iwBytes, err := iw.Marshal()
+		if err != nil {
+			return nil, fmt.Errorf("marshaling pay for blob tx: %w", err)
+		}
+		if err := pfbTxWriter.WriteTx(iwBytes); err != nil {
+			return nil, fmt.Errorf("writing pay for blob tx into compact shares: %w", err)
+		}
+	}
+
+	// defensively check that the counter is always greater in share count than the pfbTxWriter.
+	if c.pfbCounter.Size() < pfbTxWriter.Count() {
+		return nil, fmt.Errorf("pfbCounter.Size() < pfbTxWriter.Count(): %d < %d", c.pfbCounter.Size(), pfbTxWriter.Count())
+	}
+
+	// Write out the square
+	return WriteSquare(txWriter, pfbTxWriter, blobWriter, nonReservedStart, ss)
+}
+
+func (c *Builder) canFit(shareNum int) bool {
+	return c.currentSize+shareNum <= c.maxCapacity
+}
+
+func (c *Builder) isEmpty() bool {
+	return c.txCounter.Size() == 0 && c.pfbCounter.Size() == 0
+}
+
+type element struct {
+	blob       core.Blob
+	pfbIndex   int
+	blobIndex  int
+	shareSize  int
+	maxPadding int
+}
+
+func newElement(blob core.Blob, pfbIndex, blobIndex int) *element {
+	shareSize := shares.SparseSharesNeeded(uint32(len(blob.Data)))
+	return &element{
+		blob:      blob,
+		pfbIndex:  pfbIndex,
+		blobIndex: blobIndex,
+		shareSize: shareSize,
+		//
+		// For cacluating the maximum possible padding consider the following tree
+		// where each leaf corresponds to a share.
+		//
+		//	Depth       Position
+		//	0              0
+		//	              / \
+		//	             /   \
+		//	1           0     1
+		//	           /\     /\
+		//	2         0  1   2  3
+		//	         /\  /\ /\  /\
+		//	3       0 1 2 3 4 5 6 7
+		//
+		// Imagine if, according to the share commitment rules, a transcation took up 11 shares
+		// and had the merkle mountain tree commitment of 4,4,2,1. The first part of the share commitment
```

Comment:
I'm having a tough time relating this example to the ascii diagram above b/c it discusses a blob that occupies 11 shares but there are only 8 possible share indexes in the ASCII diagram

Reply:
The 8 possible shares indexes are a snapshot in the tree. It's used to demonstrate that if the index of the last transaction ended at 1 (inclusive), and the SubtreeWidth() was 4, then the next one would start at index 4 and we would have 2 shares of padding

Reply:
I copied this tree across from somewhere else just so I could explain why we have padding and why that formula gives us the maximum possible padding